### ISSUE TITLE
Read GL_R32F texture failed

### ISSUE THREAD

**Issue by 0871087123** [openscenegraph/OpenSceneGraph#664]
https://github.com/openscenegraph/OpenSceneGraph/blob/85504b45f6229c2695bc109d771045fcb10fc1e0/src/osg/Image.cpp#L425

I failed on try to read a 32bit float texture which has only one channel,  this is my code:
```
            state->applyAttribute(_texture);
            osg::ref_ptr<osg::Image> img = new osg::Image;
            img->readImageFromCurrentTexture(ri.getContextID(), true, GL_FLOAT);
```
I found a invalid enumerant error report after call glGetTexImage
I checked the code
glGetTexImage should have pixel format GL_RED as one of its' input param, but actually is GL_R32F， and it's an invalid enumerant
I found it should be computed here, from GL_R32F to GL_RED

Also , the desgin of readImageFromCurrentTexture confused me. It implemented as check every texture binding in order, it leads to if i has a cube texture binding, and currently applied a 2d texture, it will still read from cube texture. so i use this interface like this:
```
            glBindTexture(GL_TEXTURE_CUBE_MAP, 0);
            state->applyAttribute(_texture);
            osg::ref_ptr<osg::Image> img = new osg::Image;
            img->readImageFromCurrentTexture(ri.getContextID(), true, GL_FLOAT);
```
But it's not a good idea unbind cube texture when reading a 2d texture, i just use this method for test.

**Comment by openscenegraph** [comment]
I've tried to piece together what might be going on at your end, best guess at this point is that you are reading from a texture that as an internal format that computePixelFormat() isn't yet supporting so is falling back to just returning the input formatted, and the later GL call isn't handling this particular internal format.  

I'm just guessing though as I don't have a test program, any knowledge of your wider application usage, what platform you are working on, what OSG version, what hardware....

Could you create a small example that illustrates your usage case that is failing.  Taking an existing OSG example and modifying would be a good place to start.  Once we have the usage cause that is failing we can work out whether it's a usage error or an internal bug, if the later we can then fix it and use the program to confirm the issue is fixed.

Thanks.

**Comment by 0871087123** [comment]
> I've tried to piece together what might be going on at your end, best guess at this point is that you are reading from a texture that as an internal format that computePixelFormat() isn't yet supporting so is falling back to just returning the input formatted, and the later GL call isn't handling this particular internal format.
> 
> I'm just guessing though as I don't have a test program, any knowledge of your wider application usage, what platform you are working on, what OSG version, what hardware....
> 
> Could you create a small example that illustrates your usage case that is failing. Taking an existing OSG example and modifying would be a good place to start. Once we have the usage cause that is failing we can work out whether it's a usage error or an internal bug, if the later we can then fix it and use the program to confirm the issue is fixed.
> 
> Thanks.

You 're welcome. 
Thanks for reply, it's right that computePixelFormat() has some input param not supporting yet, and i'm trying to fix this by a pull request, still working on this.

**Comment by openscenegraph** [comment]
Any progress on creating a fix?

**Comment by 0871087123** [comment]
@openscenegraph I m so sorry for my late. And i will push my fix in 48 hour. Fix is not complex, but i m trying to find out is there any related problems exist. I spend a lot of time working and studying , so sorry for my late.

**Comment by openscenegraph** [comment]
I have now merged the fixes with master and the 3.6 branch so will close to Issue.
